File: 009 Palindrome Number.py
'''
Determine whether an integer is a palindrome. Do this without extra space.

Could negative integers be palindromes? (ie, -1)

If you are thinking of converting the integer to string, note the restriction of using extra space.

You could also try reversing an integer. However, if you have solved the problem "Reverse Integer", you know that the reversed integer might overflow. How would you handle such case?

There is a more generic way of solving this problem.
'''

class Solution(object):
    def isPalindrome(self, x):
        """
        :type x: int
        :rtype: bool
        """
        if x < 0:
            return False
        div = 1
        while x / div >= 10:
            div *= 10
        while x > 0:
            l = x // div
            r = x % 10

            if l != r:
                return False
            x %= div
            x //= 10
            div /= 100
        return True


if __name__ == "__main__":
    assert Solution().isPalindrome(123) == False
    assert Solution().isPalindrome(12321) == True
    assert Solution().isPalindrome(-121) == False

File: java/_009PalindromeNumber.java
/**
 * Determine whether an integer is a palindrome. Do this without extra space.
 * <p>
 * Some hints:
 * Could negative integers be palindromes? (ie, -1)
 * <p>
 * If you are thinking of converting the integer to string, note the restriction of using extra space.
 * <p>
 * You could also try reversing an integer. However, if you have solved the problem "Reverse Integer", you know that
 * the reversed integer might overflow. How would you handle such case?
 * <p>
 * There is a more generic way of solving this problem.
 * <p>
 * Created by drfish on 25/06/2017.
 */
public class _009PalindromeNumber {
    public boolean isPalindrome(int x) {
        if (x < 0 || (x != 0 && x % 10 == 0)) {
            return false;
        }
        int reverse = 0;
        while (x > reverse) {
            reverse = reverse * 10 + x % 10;
            x /= 10;
        }
        return (x == reverse || x == reverse / 10);
    }

    public static void main(String[] args) {
        _009PalindromeNumber solution = new _009PalindromeNumber();
        assert solution.isPalindrome(12321);
        assert solution.isPalindrome(123321);
        assert !solution.isPalindrome(-121);
    }
}
